lib/fs.js

```diff
@@ -395,12 +395,32 @@ function readFileAfterClose(err) {
   else
     buffer = context.buffer;
 
-  if (context.encoding)
-    buffer = buffer.toString(context.encoding);
+  if (err) return callback(err, buffer);
 
-  callback(err, buffer);
+  if (context.encoding) {
+    const result = tryToString(buffer, context.encoding);
+    if (result.error)
+      return callback(result.error, buffer);
```

Comment:
Yea, that is why I did that. And honestly, I feel like this is something that should go into v4.x if possible. We could make it better for v5?